[Task] Loading JFR event data blocks the Quarkus event thread

```
2021-11-12 22:42:01,407 INFO  [io.cry.jfr.dat.eve.RecordingService] (vert.x-eventloop-thread-6) Loading file: /tmp/jfr-file-uploads/vertx-fib-demo-6f4775cdbf-82dvl_150mb_20211006T152006Z.jfr
2021-11-12 22:42:05,336 WARN  [io.ver.cor.imp.BlockedThreadChecker] (vertx-blocked-thread-checker) Thread Thread[vert.x-eventloop-thread-6,5,main]=Thread[vert.x-eventloop-thread-6,5,main] has been blocked for 3928 ms, time limit is 2000 ms: io.vertx.core.VertxException: Thread blocked
	at com.oracle.svm.core.posix.headers.Pthread.pthread_cond_wait(Pthread.java)
	at com.oracle.svm.core.posix.thread.PosixParkEvent.condWait(PosixJavaThreads.java:250)
	at com.oracle.svm.core.thread.JavaThreads.park(JavaThreads.java:688)
	at jdk.internal.misc.Unsafe.park(Unsafe.java:49)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:194)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2081)
	at java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:433)
	at java.util.concurrent.ExecutorCompletionService.take(ExecutorCompletionService.java:200)
	at org.openjdk.jmc.flightrecorder.internal.FlightRecordingLoader.readChunks(FlightRecordingLoader.java:242)
	at org.openjdk.jmc.flightrecorder.internal.FlightRecordingLoader.loadStream(FlightRecordingLoader.java:96)
	at org.openjdk.jmc.flightrecorder.JfrLoaderToolkit.loadFile(JfrLoaderToolkit.java:71)
	at org.openjdk.jmc.flightrecorder.JfrLoaderToolkit.loadEvents(JfrLoaderToolkit.java:143)
	at org.openjdk.jmc.flightrecorder.JfrLoaderToolkit.loadEvents(JfrLoaderToolkit.java:128)
	at org.openjdk.jmc.flightrecorder.JfrLoaderToolkit.loadEvents(JfrLoaderToolkit.java:116)
	at io.cryostat.jfr.datasource.events.RecordingService.loadEvents(RecordingService.java:225)
	at io.cryostat.jfr.datasource.events.RecordingService_ClientProxy.loadEvents(RecordingService_ClientProxy.zig:129)
	at io.cryostat.jfr.datasource.server.JfrResource.setFile(JfrResource.java:162)
	at io.cryostat.jfr.datasource.server.JfrResource.load(JfrResource.java:103)
	at io.cryostat.jfr.datasource.server.JfrResource_RouteHandler_load_3c03e47a7c388a16485114470151eb9b146bef6a.invoke(JfrResource_RouteHandler_load_3c03e47a7c388a16485114470151eb9b146bef6a.zig:107)
	at io.quarkus.vertx.web.runtime.RouteHandler.handle(RouteHandler.java:97)
	at io.quarkus.vertx.web.runtime.RouteHandler.handle(RouteHandler.java:22)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1038)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:101)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:132)
	at io.vertx.ext.web.handler.impl.BodyHandlerImpl$BHandler.doEnd(BodyHandlerImpl.java:296)
	at io.vertx.ext.web.handler.impl.BodyHandlerImpl$BHandler.uploadEnded(BodyHandlerImpl.java:265)
	at io.vertx.ext.web.handler.impl.BodyHandlerImpl$BHandler.lambda$null$1(BodyHandlerImpl.java:204)
	at io.vertx.core.http.impl.HttpServerFileUploadImpl.handleEnd(HttpServerFileUploadImpl.java:87)
	at io.vertx.core.http.impl.HttpServerFileUploadImpl.lambda$null$1(HttpServerFileUploadImpl.java:178)
	at io.vertx.core.impl.ContextImpl.lambda$null$0(ContextImpl.java:327)
	at io.vertx.core.impl.ContextImpl.executeTask(ContextImpl.java:366)
	at io.vertx.core.impl.EventLoopContext.lambda$executeAsync$0(EventLoopContext.java:38)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:834)
	at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:519)
	at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:192)
```

This work should be done asynchronously so that the event thread is not blocked.

@andrewazores I was able to recreate and have an initial fix so that the main thread is not blocked. But it notice there is a timeout that throws an exception if IO operation takes too long (>29s) to return a response. Is that some hard limit we need to follow? I am not sure how we go about this issue?

```
SEVERE: HTTP 500: io.vertx.core.http.impl.NoStackTraceTimeoutException: The timeout period of 29000ms has been exceeded while executing POST /load for server localhost:8080
io.vertx.ext.web.handler.impl.HttpStatusException: Internal Server Error
Caused by: java.util.concurrent.ExecutionException: io.vertx.core.http.impl.NoStackTraceTimeoutException: The timeout period of 29000ms has been exceeded while executing POST /load for server localhost:8080
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:395)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1999)
	at io.cryostat.net.web.http.api.v1.RecordingUploadPostHandler.doPost(RecordingUploadPostHandler.java:192)
	at io.cryostat.net.web.http.api.v1.RecordingUploadPostHandler.handleAuthenticated(RecordingUploadPostHandler.java:137)
	at io.cryostat.net.web.http.AbstractAuthenticatedRequestHandler.handle(AbstractAuthenticatedRequestHandler.java:99)
	at io.cryostat.net.web.http.AbstractAuthenticatedRequestHandler.handle(AbstractAuthenticatedRequestHandler.java:69)
	at io.vertx.ext.web.impl.BlockingHandlerDecorator.lambda$handle$0(BlockingHandlerDecorator.java:48)
	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$2(ContextImpl.java:313)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: io.vertx.core.http.impl.NoStackTraceTimeoutException: The timeout period of 29000ms has been exceeded while executing POST /load for server localhost:8080
```

--- a/src/main/java/io/cryostat/jfr/datasource/server/JfrResource.java
+++ b/src/main/java/io/cryostat/jfr/datasource/server/JfrResource.java
@@ -91,14 +91,16 @@ public class JfrResource {
     HttpServerResponse response = context.response();
     setHeaders(response);
 
-    StringBuilder responseBuilder = new StringBuilder();
+    final StringBuilder responseBuilder = new StringBuilder();
     
     context.vertx().executeBlocking(
       future -> {
-      uploadFiles(context.fileUploads(), responseBuilder);
-      future.complete();
+        uploadFiles(context.fileUploads(), responseBuilder);
+        future.complete();
     }, false, 
-      result -> response.end(responseBuilder.toString())
+      result -> {
+        response.end(responseBuilder.toString());
+      }
     );
   }
 

--- a/src/main/java/io/cryostat/jfr/datasource/server/JfrResource.java
+++ b/src/main/java/io/cryostat/jfr/datasource/server/JfrResource.java
@@ -14,6 +14,7 @@ import org.eclipse.microprofile.config.inject.ConfigProperty;
 
 import io.cryostat.jfr.datasource.events.RecordingService;
 import io.quarkus.vertx.web.Route;
+import io.smallrye.common.annotation.Blocking;
 import io.vertx.core.http.HttpServerResponse;
 import io.vertx.core.json.JsonObject;
 import io.vertx.core.logging.Logger;
@@ -70,6 +71,7 @@ public class JfrResource {
   }
 
   @Route(path = "/set")
+  @Blocking
   void set(RoutingContext context) {
     HttpServerResponse response = context.response();
     setHeaders(response);
@@ -77,57 +79,33 @@ public class JfrResource {
     String file = context.getBodyAsString();
     String filePath = jfrDir + File.separator + file;
 
-    context.vertx().executeBlocking(
-      future -> {
-        future.complete();
-        setFile(filePath, file, response, new StringBuilder());
-      }, false,
-      result -> {}
-    );
+    setFile(filePath, file, response, new StringBuilder());
   }
 
   @Route(path = "/upload")
+  @Blocking
   void upload(RoutingContext context) {
     HttpServerResponse response = context.response();
     setHeaders(response);
 
     final StringBuilder responseBuilder = new StringBuilder();
-    
-    context.vertx().executeBlocking(
-      future -> {
-        uploadFiles(context.fileUploads(), responseBuilder);
-        future.complete();
-    }, false, 
-      result -> {
-        response.end(responseBuilder.toString());
-      }
-    );
+
+    uploadFiles(context.fileUploads(), responseBuilder);
+    response.end(responseBuilder.toString());
   }
 
   @Route(path = "/load")
+  @Blocking
   void load(RoutingContext context) {
     HttpServerResponse response = context.response();
     setHeaders(response);
 
     final StringBuilder responseBuilder = new StringBuilder();
 
-    context.vertx().executeBlocking(
-      future -> {
-        String lastFile = uploadFiles(context.fileUploads(), responseBuilder);
-        future.complete(lastFile);
-      }, false,
-      result -> {
-        String lastFile = (String) result.result();
-        String filePath = jfrDir + File.separator + lastFile;
-        context.vertx().executeBlocking(
-          nextFuture -> {
-            nextFuture.complete();
-            setFile(filePath, lastFile, response, responseBuilder);
-          }, false,
-          nextResult -> {}
-        );
-      }
-    );
+    String lastFile = uploadFiles(context.fileUploads(), responseBuilder);
+    String filePath = jfrDir + File.separator + lastFile;
+
+    setFile(filePath, lastFile, response, responseBuilder);
   }
 
   @Route(path = "/list")

--- a/src/main/java/io/cryostat/jfr/datasource/server/JfrResource.java
+++ b/src/main/java/io/cryostat/jfr/datasource/server/JfrResource.java
@@ -77,7 +77,13 @@ public class JfrResource {
     String file = context.getBodyAsString();
     String filePath = jfrDir + File.separator + file;
 
-    setFile(filePath, file, response, new StringBuilder());
+    context.vertx().executeBlocking(
+      future -> {
+        future.complete();
+        setFile(filePath, file, response, new StringBuilder());
+      }, false,
+      result -> {}
+    );
   }
 
   @Route(path = "/upload")
@@ -86,9 +92,14 @@ public class JfrResource {
     setHeaders(response);
 
     StringBuilder responseBuilder = new StringBuilder();
-    uploadFiles(context.fileUploads(), responseBuilder);
-
-    response.end(responseBuilder.toString());
+    
+    context.vertx().executeBlocking(
+      future -> {
+      uploadFiles(context.fileUploads(), responseBuilder);
+      future.complete();
+    }, false, 
+      result -> response.end(responseBuilder.toString())
+    );
   }
 
   @Route(path = "/load")
@@ -96,11 +107,25 @@ public class JfrResource {
     HttpServerResponse response = context.response();
     setHeaders(response);
 
-    StringBuilder responseBuilder = new StringBuilder();
-    String lastFile = uploadFiles(context.fileUploads(), responseBuilder);
-    String filePath = jfrDir + File.separator + lastFile;
+    final StringBuilder responseBuilder = new StringBuilder();
 
-    setFile(filePath, lastFile, response, responseBuilder);
+    context.vertx().executeBlocking(
+      future -> {
+        String lastFile = uploadFiles(context.fileUploads(), responseBuilder);
+        future.complete(lastFile);
+      }, false,
+      result -> {
+        String lastFile = (String) result.result();
+        String filePath = jfrDir + File.separator + lastFile;
+        context.vertx().executeBlocking(
+          nextFuture -> {
+            nextFuture.complete();
+            setFile(filePath, lastFile, response, responseBuilder);
+          }, false,
+          nextResult -> {}
+        );
+      }
+    );
   }
 
   @Route(path = "/list")

